shesha/supervisor/saxoPlusManager.py

```diff
@@ -116,37 +120,77 @@ def __init__(self, first_stage, second_stage):
         self.second_stage_input = np.zeros((mpup_shape[0], mpup_shape[1], 1))
         residual_shape = self.first_stage.config.p_geom._spupil.shape
         self.mpup_offset = (mpup_shape[0] - residual_shape[0]) // 2
-        self.frequency_ratio = round(1e-3 / self.second_stage.config.p_loop.ittime)
+
+        # Must be an input to configure the manager
+        self.frequency_ratio = frequency_ratio
 
         # flags for enabling asterix coronagraph
         self.first_stage.computeCoroImage = False
         self.second_stage.computeCoroImage = False
 
-    def next(self):
+    def next(self, *, do_control: bool = True) -> None:
         """
         MAIN method that allows to manage properly the 2 AO stages of SAXO+ system. 
         The phase residuals (including turbulence + AO loop residuals) of the first stage simulation is sent to second stage simulation
         at each iteration of the manager. 
         The saxo+ manager disable the seconds stage turbulence simulation (as it is propageated through the first stage residals if any). 
 
         This next method sould ALWAYS be called to perform a regular SAXO+ simulation 
-        instead of the individuals COMPASS next methods to ensure the correct synchronisation of the 2 systems. 
+        instead of the individual stage next methods to ensure the correct synchronisation of the 2 systems. 
         """
-        # Iteration time of the first stage is set as the same as the second stage to allow
-        # correct atmosphere movement for second stage integration. Then, first stage as to compute
-        # every 3 iteration to be 3 times slower than the second stage
-        self.second_stage.atmos.enable_atmos(False) # Turbulence always disabled on 2nd instance of COMPASS
-        if not (self.iterations % self.frequency_ratio): # Time for first stage full computation: We update the first stage command every frequency_ratio iterations. 
-            self.first_stage.next()
-        else: # Only raytracing current tubulence phase (if any) and current DMs phase (no command updates). 
-            self.first_stage.next(do_control=False, apply_control=False, compute_tar_psf=True)
+        # Iteration time of the first stage is set as the same as the second stage to
+        # allow correct atmosphere movement for second stage integration. Then,
+        # first stage is controlled only once every frequency_ratio times
+
+        # Turbulence always disabled on 2nd instance of COMPASS                         
+        self.second_stage.atmos.enable_atmos(False) 
+        
+        if do_control:
+            # system has already been calibrated and normal run can be done
+
+            if not (self.iterations % self.frequency_ratio):
```

Comment:
I think there is an issue if `frequency_ratio = 1`. In that case, control is never done. Maybe we can with the 'if not ((self.iterations+1) % self.frequency_ratio)' lines ?

Reply:
You are right, there is an issue in the case frequency_ratio =1. And switching the "if" is not sufficient to solve it. I will remove the 3 if/elif/else here and define flags for do_control, do_centroids and stack_wfs_image in advance, in order to call only once the next function with the proper combination of keywords values whatever the integer frequency_ratio.

Reply:
I modified the manager and checked that the system still works (providing correct calibrations and correct Strehl ratios at the two stages).